### ISSUE TITLE
IMAP4 server loops indefinitely if a search query yields more than 4 items

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4385]
|<img alt="wulczer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wulczer reported|
|-|-|
|Trac ID|trac#4385|
|Type|defect|
|Created|2010-03-29 20:22:39Z|

The responsible code seems to be
```
if i == 4:
    from twisted.internet import reactor
    reactor.callLater(
        0, self.__cbManualSearch, result, tag, mbox, query, uid,
        searchResults)
```
in IMAP4Server.__cbManualSearch.

It seems to me that it should be slicing the first four items off the result list before calling itself again.

Attachments:

* [twisted-4385-imap4-loop.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/013/013e8477c07fa07eaa1f44db057a46374a1579aa/162e3efd226c4fc7e87f99f2232d3a727cc562c6.patch) (2916 bytes) - added by wulczer on 2010-03-29 20:30:20Z - patch to fix the issue
* [twisted-4385-imap4-loop-v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/013/013e8477c07fa07eaa1f44db057a46374a1579aa/0284fd8260e9095195aa0d6f2d4a1c3006818cd6.patch) (4643 bytes) - added by wulczer on 2010-04-09 19:52:54Z - rebased patch to trunk

<details><summary>Searchable metadata</summary>

```
trac-id__4385 4385
type__defect defect
reporter__wulczer wulczer
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1269894159000000 1269894159000000
changetime__1270910102000000 1270910102000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="wulczer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|wulczer removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to wulczer|
|-|-|

`DefaultSearchTestCase.test_searchAndMessageSet` is freezing for me. The patch doesn't seem to be again latest trunk, maybe it's the problem?

**Comment by twisted-trac** [comment]
|<img alt="wulczer's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|wulczer removed owner|
|-|-|

Replying to [therve](#note_2):
> `DefaultSearchTestCase.test_searchAndMessageSet` is freezing for me. The patch doesn't seem to be again latest trunk, maybe it's the problem?

Yes, you're right, the patch was bad. I rebased it against latest trunk (attached).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks.  The patch looks good.  I'll apply it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [28818]) Apply twisted-4385-imap4-loop-v2.patch to fix the infinite loop in IMAP4 server search handling

Author: wulczer
Reviewer: therve, exarkun
Fixes: [#4385](https://github.com/twisted/twisted/issues/4385)

Previously the fallback search implementation provided by IMAP4Server would examine the first five
messages in a mailbox forever, trying to respond to a search command.  This fixes it to actually
iterate over all the messages in the mailbox and eventually complete.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
